webgoat-lessons/csrf/src/main/java/org/owasp/webgoat/csrf/ForgedReviews.java

```diff
@@ -0,0 +1,107 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.MediaType;
+import org.springframework.web.bind.annotation.*;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.*;
+
+import static org.springframework.http.MediaType.ALL_VALUE;
+
+@RestController
+@AssignmentHints({"csrf-review-hint1", "csrf-review-hint2", "csrf-review-hint3"})
+public class ForgedReviews extends AssignmentEndpoint {
+
+    @Autowired
+    private WebSession webSession;
+    private static DateTimeFormatter fmt = DateTimeFormat.forPattern("yyyy-MM-dd, HH:mm:ss");
+
+    private static final Map<String, List<Review>> userReviews = new HashMap<>();
+    private static final List<Review> REVIEWS = new ArrayList<>();
+    private static final String weakAntiCSRF = "2aa14227b9a13d0bede0388a7fba9aa9";
+
+
+    static {
+        REVIEWS.add(new Review("secUriTy", DateTime.now().toString(fmt), "This is like swiss cheese", 0));
+        REVIEWS.add(new Review("webgoat", DateTime.now().toString(fmt), "It works, sorta", 2));
+        REVIEWS.add(new Review("guest", DateTime.now().toString(fmt), "Best, App, Ever", 5));
+        REVIEWS.add(new Review("guest", DateTime.now().toString(fmt), "This app is so insecure, I didn't even post this review, can you pull that off too?", 1));
+    }
+
+    @GetMapping(path = "/csrf/review", produces = MediaType.APPLICATION_JSON_VALUE, consumes = ALL_VALUE)
+    @ResponseBody
+    public Collection<Review> retrieveReviews() {
+        Collection<Review> allReviews = Lists.newArrayList();
+        Collection<Review> newReviews = userReviews.get(webSession.getUserName());
+        if (newReviews != null) {
+            allReviews.addAll(newReviews);
+        }
+
+        allReviews.addAll(REVIEWS);
+
+        return allReviews;
+    }
+
+    @PostMapping("/csrf/review")
+    @ResponseBody
+    public AttackResult createNewReview(String reviewText, Integer stars, String validateReq, HttpServletRequest request) {
+        final String host = (request.getHeader("host") == null) ? "NULL" : request.getHeader("host");
+        final String referer = (request.getHeader("referer") == null) ? "NULL" : request.getHeader("referer");
+        final String[] refererArr = referer.split("/");
+
+        Review review = new Review();
+        review.setText(reviewText);
+        review.setDateTime(DateTime.now().toString(fmt));
+        review.setUser(webSession.getUserName());
+        review.setStars(stars);
+        var reviews = userReviews.getOrDefault(webSession.getUserName(), new ArrayList<>());
+        reviews.add(review);
+        userReviews.put(webSession.getUserName(), reviews);
+        //short-circuit
+        if (validateReq == null || !validateReq.equals(weakAntiCSRF)) {
+            return failed(this).feedback("csrf-you-forgot-something").build();
+        }
+        //we have the spoofed files
+        if (referer != "NULL" && refererArr[2].equals(host)) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.